include/deal.II/lac/qr.h

```diff
@@ -0,0 +1,860 @@
+
+#include <boost/signals2/signal.hpp>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A base class for QR implementations.
+ */
+template <typename VectorType>
+class BaseQR
+{
+  /**
+   * Number type for R matrix.
+   */
+  typedef typename VectorType::value_type Number;
+
+protected:
+  /**
+   * Default private constructor.
+   */
+  BaseQR();
+
+public:
+  /**
+   * Destructor.
+   */
+  virtual ~BaseQR() = default;
+
+  /**
+   * Append @p column to the QR factorization.
+   * Returns <code>true</code> if the result is successful, i.e.
+   * the columns are linearly independent. Otherwise the @p column
+   * is rejected and the return value is <code>false</code>.
+   */
+  virtual bool
+  append_column(const VectorType &column) = 0;
+
+  /**
+   * Return size of the subspace.
+   */
+  unsigned int
+  size() const;
+
+  /**
+   * Remove a column @p k and update QR factorization.
+   */
+  virtual void
+  remove_column(const unsigned int k = 0) = 0;
+
+  /**
+   * Return the current upper triangular matrix R.
+   */
+  const LAPACKFullMatrix<Number> &
+  get_R() const;
+
+  /**
+   * Solve $Rx=y$. Vectors @p x and @p y should be consistent
+   * with the current size of the subspace.
+   * If @p transpose is <code>true</code>, $R^Tx=y$ is solved instead.
+   */
+  void
+  solve(Vector<Number> &      x,
+        const Vector<Number> &y,
+        const bool            transpose = false) const;
+
+  /**
+   * Set $y = Q*x$. The size of $x$ should be consistent with the
+   * size of the R matrix.
+   */
+  virtual void
+  multiply_with_Q(VectorType &y, const Vector<Number> &x) const = 0;
+
+  /**
+   * Set $y = Q^Tx$. The size of $x$ should be consistent with the
+   * size of column vectors.
+   */
+  virtual void
+  multiply_with_QT(Vector<Number> &y, const VectorType &x) const = 0;
+
+  /**
+   * Set $y = Q*R*x$. The size of $x$ should be consistent with the
+   * size of the R matrix.
+   */
+  virtual void
+  multiply_with_A(VectorType &y, const Vector<Number> &x) const = 0;
+
+  /**
+   * Set $y = R^T Q^Tx$. The size of $x$ should be consistent with the
+   * size of column vectors.
+   */
+  virtual void
+  multiply_with_AT(Vector<Number> &y, const VectorType &x) const = 0;
+
+  /**
+   * Connect a slot to retrieve a notification when the Givens rotations
+   * are performed.
+   */
+  boost::signals2::connection
+  connect_givens_slot(
+    const std::function<void(const unsigned int i,
+                             const unsigned int j,
+                             const std::array<Number, 3> &)> &slot);
+
+protected:
+  /**
+   * Multiply with columns stored in the object.
+   */
+  void
+  multiply_with_cols(VectorType &y, const Vector<Number> &x) const;
+
+  /**
+   * Multiply with transpose columns stored in the object.
+   */
+  void
+  multiply_with_colsT(Vector<Number> &y, const VectorType &x) const;
+
+  /**
+   * A vector of shared pointers to store columns.
+   */
+  std::vector<std::shared_ptr<VectorType>> columns;
```

Comment:
they are shared in order to be able to return `const` shared pointers outside.

I am not sure If there is need for caching similar to https://www.dealii.org/developer/doxygen/deal.II/classFiniteSizeHistory.html

Reply:
What I meant to say is this: Having a `std::shared_ptr` only makes sense if there are other places in the code that want to share ownership of a vector. If all you want to do is let others *see* a vector, then you can always return a const reference to it. It's really only necessary to let others share ownership if they will want to continue using the vector *after* the class that has the member variable is already gone. Is this what you want to do? Can you give an example how a caller of the function that returns the shared pointer will want to keep working on these vectors after the QR class is already gone?

Reply:
> If all you want to do is let others see a vector, then you can always return a const reference to it.

you mean 
```
const std::unique_ptr<VectorType> &
```
?

> Is this what you want to do? 

nope.

Reply:
Yes, a `std::vector<std::unique_ptr>` would work.

Or maybe I just don't understand your use case. Can you show me the piece of code where you want to call the function that returns the `std::vector` of shared pointers?

Reply:
> Can you show me the piece of code where you want to call the function that returns the std::vector of shared pointers?

IIRC it's only for tests, the actual usage is always through the interface of the class, i.e. `multiply_with_XYZ`.

Reply:
So the ownership of the vectors is unambiguous: This class owns them, and no other class will use them other than through `const` references. In that case, I'd store `std::vector<std::unique_ptr>` and either return a `const` reference to that object, or `const` references to the individual vectors if necessary.

`shared_ptr` only makes sense if another class might want to know about these vectors and potentially take over ownership of them because they are still used after the current `BaseQR` object dies. But I don't think this is your use case.